# 
$${\color{red}DAY 3}$$
# 

$\color{#58A6FF}\textsf{\Large\&#x24D8;\kern{0.2cm}\normalsize Note}$
Do not forget to activate the conda environment

``` 
module load gcc12-env/12.1.0
module load miniconda3/4.12.0
conda activate anvio-8
``` 
## Assembly visualization

To count the number of contigs in your final assembly file
```
grep -c ">" final.contigs.fa
```

To visualize contig graph in **Bandage**, the first step is to convert the fasta file(s) intermediate_contigs/k{kmer_size}.contigs.fa into SPAdes-like **FASTG** format. The following code shows the translation from **NAME.contigs.fa** into **NAME.fastg**.
  
Example:
```
megahit_toolkit contig2fastg 99 final.contigs.fa > final.contigs.fastg                   
```
Then the FASTG file k99.fastg can be loaded into **Bandage**.

The program is already installed on your PC and can be used as a GUI.
Use it to open the **final.contigs.fastg** file. Once it's loaded (this might take a second) click on Draw graph to create a graph. 
  
You can now label the nodes by adding parameters like Depth or Name. Whenever you change something you need to click draw graph again.

## Questions
  
* **Please submit your generated figure and explain in your own words what you can see (keep it short).**
 
> INSERT\
> YOUR\
> ANSWER\
> HERE


## Quality Assessment of Assemblies

[quast](https://quast.sourceforge.net/quast ) is a **QU**ality **AS**sessment **T**ool to evaluate genome assembly. You will use it to evaluate the results coming from [megahit](https://github.com/voutcn/megahit ).\
First you need to run quast. Using the terminal go to the quast folder, then use the following command:

```ssh
metaquast -t 6 -o ? -m 1000 ?
```

<details><summary><b>Finished commands</b></summary>

```ssh
metaquast -t 6 -o /PATH/TO/3_metaquast -m 1000 final.contigs.fa
```
</details>

> -o path to output directory. Output is given both as a PDF and as a html file.

### Questions

* **What is your N50 value? Why is this value relevant?**
* **How many contigs are assembled?**
* **What is the total length of the contigs?**

> INSERT\
> YOUR\
> ANSWER\
> HERE

## Genomes Binning
The first thing you will do is format your fasta sequence IDs. Anvi’o (which you will use later) needs this step to work properly.\
Run this with your contigs and with your clean reads (Hint: fastp output). As the names get changed, you need to run it on your assembly file, otherwise the names of the contigs won't match the names of the initial reads (essential for the mapping step below).

```ssh
anvi-script-reformat-fasta ? -o ? --min-len 1000 --simplify-names --report-file name_conversion.txt
```

<details><summary><b>Finished commands</b></summary>

```ssh
anvi-script-reformat-fasta final.contigs.fa -o /PATH/TO/YOUR/contigs.anvio.fa --min-len 1000 --simplify-names --report-file name_conversion.txt
```
</details>

$\color{#58A6FF}\textsf{\Large\&#x24D8;\kern{0.2cm}\normalsize Note}$
If you use the flag --report-file, it will also create a TAB-delimited file for you to keep track of which defline in the new file corresponds to which defline in the original file

### Mapping

Then you need to map your raw reads onto your assembled contigs.
Mapping will be done using [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml ). Use the following command to index your mapping reference fasta file. Needed for the next steps and basically makes mapping faser.

```
module load bowtie2
bowtie2-build contigs.anvio.fa contigs.anvio.fa.index
```

> `input_file` specify the path to your contigs as input_file (for example:contigs.fa)\
> `index` index that will appear in front of all output files generated by this command (in this example your files will be named out_index.1.bt2, out_index.2.bt2 and so on, in total you should get 6 index files).

Now use [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml ) for the actual mapping. Use the loop command, below you have the single command for better understanding.

```ssh
module load bowtie2
bowtie2 --very-fast -x contigs.anvio.fa.index -1 ? -2 ? -S ?
```

<details><summary><b>Finished commands</b></summary>

```ssh
module load bowtie2
bowtie2 --very-fast -x contigs.anvio.fa.index -1 /PATH/TO/sample1_R1_clean.fastq.gz -2 /PATH/TO/sample1_R2_clean.fastq.gz -S SAMPLE.sam
```

or in a loop from your fastq clean folder: 

```ssh
module load bowtie2
cd /PATH/TO/FASTP
for i in `ls *_R1.fastq.gz`;
do
    second=`echo ${i} | sed 's/_R1/_R2/g'`
bowtie2 --very-fast -x /PATH/TO/index -1 ${i} -2 ${second} -S /PATH/TO/4_mapping/"$i".sam 
done
```
</details>

> `--very-fast` bowtie runs in very fast but less accurate end-to-end mode\
> `-x` index files with the contigs from the step before, give it the prefix name of the files (the part that comes before the dot)
 
> `-1` R1 fasta file containing the raw reads after fastp processing\
> `-2` R2 fasta file containing the raw reads after fastp processing\
> `-S` name of the output file, don't forget the .sam part!

$\color{#58A6FF}\textsf{\Large\&#x24D8;\kern{0.2cm}\normalsize Note}$
All `*.final.contigs.fasta` should have a corresponding mapping file !!!!

The output will be a sequence mapping file (SAM) with the .sam extension and which we convert to binary alignment and map (BAM) file with the .bam extension using [samtools](https://github.com/samtools/samtools ) with the following loop:

```ssh
module load samtools
samtools view -bS ? > bam_file.bam
```

<details><summary><b>Finished commands</b></summary>

```ssh
module load samtools
samtools view -bS sam_file.sam > bam_file.bam
```

Or in a loop:

```ssh
module load samtools

cd /PATH/TO/MAPPING/OUT
for i in *.sam; do samtools view -bS $i > "$i".bam; done
```
</details>

> `sam_file.sam` input .sam file 
> `bam_file.bam` output .bam file

### Contigs data preparation!
You need to convqqase is an *anvi’o* [contigs-db](https://anvio.org/help/main/artifacts/contigs-db/) database that contains key information associated with your sequences.

```
anvi-gen-contigs-database -f contigs.anvio.fa -o contigs.db -n 'biol217'
```

> `-f` contig.fa files, used as input\
> `-o` will give you a .db file as output

When you run this command, ``anvi-gen-contigs-database`` will (documentation [anvi-gen-contigs-database](https://merenlab.org/2016/06/22/anvio-tutorial-v2/#creating-an-anvio-contigs-database ))

* Compute k-mer frequencies for each contig 
* Soft-split contigs longer than 20,000 bp into smaller ones
* Identify open reading frames using Prodigal, the bacterial and archaeal gene finding program

Then you need to perform an HMM search on your contigs. *"Basically, in [anvi’o](https://anvio.org/ ), Hidden Markov Models (or HMMs for short) are used to search for specific genes with known functions in a larger dataset"* (documentation [anvi-run-hmms] (https://anvio.org/help/7/programs/anvi-run-hmms/)

```
anvi-run-hmms -c contigs.db
```
When you run this command (without any other parameters),
* It will utilize multiple default bacterial single-copy core gene collections and identify hits among your genes to those collections
* Note that the program will use only one CPU by default, especially if you have multiple of them available to you, you should use the ``--num-threads`` parameter.

Once you have your contigs database ready, and optionally your HMMs are run, you can take a quick look at it using the program ``anvi-display-contigs-stats``:


```diff
- Here you need to access anvi’o interactive -
- REPLACE the command line you want to run in interactive mode -
```

```
module load gcc12-env/12.1.0
module load miniconda3/4.12.0
conda activate anvio-8

anvi-display-contigs-stats contigs.db
```

This program shows you simple stats of your contigs database that may help you not only assess your assembly output, but also estimate the number of bacterial and archaeal genomes to recover.

### Binning with ANVI´O 

From here on you will be using **ANVI´O**, an **AN**alysis and **Visualization** platform for microbial **´O**mics.

*“[anvi’o](https://anvio.org/ ) is a comprehensive platform that brings together many aspects of today’s cutting-edge computational strategies of data-enabled microbiology, including genomics, metagenomics, metatranscriptomics, pangenomics, metapangenomics, phylogenomics, and microbial population genetics in an integrated and easy-to-use fashion through extensive interactive visualization capabilities.”*

Together with a very extensive program, [anvi’o](https://anvio.org/ ) offers a lot of explanations on every command or data type it uses, so, whenever you need more information than provided here, look through the online tutorial [Anvi'o User Tutorial for Metagenomic Workflow](https://merenlab.org/2016/06/22/anvio-tutorial-v2/ ) !

#### Sort and Index bam files

The first step will be to sort and index your ``.bam`` files.

[anvi’o](https://anvio.org/ ) does this by using [samtools](https://github.com/samtools/samtools ) in the background, it merges two separate samtools commands (sorting and indexing = fo each ``.bam`` file you have, there also is a ``.bam.bai`` file in the same directory) into one. 

Run the loop as shown below:

```
for i in *.bam; do anvi-init-bam $i -o "$i".sorted.bam; done
```


Now you can use [anvi’o](https://anvio.org/ ) to perform genome binning. In this step you will group your contigs and assign them to individual genomes. As with [samtools](https://github.com/samtools/samtools ), [anvi’o](https://anvio.org/ )  uses binners like [Metabat2](https://bitbucket.org/berkeleylab/metabat/src/master/ ), [binsanity](https://github.com/edgraham/BinSanity) or other binners in the background. 

A few preparation steps are required before you perform the actual binning. 

#### Creating an Anvi’o Profile

An [anvi’o](https://anvio.org/ ) profile stores sample-specific information about contigs. Profiling a BAM file with anvi’o using ``anvi-profile`` creates a single profile that reports properties for each contig in a single sample based on mapping results.

```ssh
anvi-profile -i ? -c ? --output-dir ?
```

<details><summary><b>Finished commands</b></summary>

```ssh
anvi-profile -i YOUR_SORTED.bam -c contigs.db --output-dir OUTPUT_DIR
```

or in a loop:

```ssh
mkdir /PATH/TO/profiling/
for i in `ls *.sorted.bam | cut -d "." -f 1`; do anvi-profile -i "$i".bam.sorted.bam -c contigs.db -o /PATH/TO/profiling/”$i”; done
```
</details>

> `-i` sorted and indexed bam file, used as input\
> `-c` contig.db file\
> `--output-dir` give your output directory a name

When you run ``anvi-profile`` [profile-db](https://anvio.org/help/main/artifacts/profile-db/) it will:

* Process each contig that is longer than **2,500 nts** by default. Remember that the minimum contig length should be long enough for tetra-nucleotide frequencies to have enough meaningful signal. Never go below 1,000 nts.

Processing of contigs will include:

* The recovery of mean coverage, standard deviation of coverage, and the average coverage for the inner quartiles (Q1 and Q3) for a given contig.
* The characterization of single-nucleotide variants (SNVs) for every nucleotide position. By default, the profiler will not pay attention to any nucleotide position with less than 10X coverage. 
* Finally, because single profiles are rarely used for genome binning or visualization, and since the clustering step increases the profiling runtime for no good reason, the default behavior of profiling is to not cluster contigs automatically.

This command line will return a folder where you will find the following files:

> `RUNNLOG.txt` the anvi´o log of your run\
> `PROFILE.db` An anvi’o database that contains key information about the mapping of short reads from multiple samples to your contigs 

make sure the output folders don't have "". if so then please rename it!!!
In the next step, you will merge the profiles coming from your different samples into one profile:

```ssh
anvi-merge /PATH/TO/SAMPLE1/? /PATH/TO/SAMPLE2/? /PATH/TO/SAMPLE3/? -o ? -c ? --enforce-hierarchical-clustering
```

<details><summary><b>Finished commands</b></summary>

```ssh
anvi-merge /PATH/TO/SAMPLE1/PROFILE.db /PATH/TO/SAMPLE2/PROFILE.db /PATH/TO/SAMPLE3/PROFILE.db -o /PATH/TO/merged_profiles -c /PATH/TO/contigs.db --enforce-hierarchical-clustering
```
</details>

> `profile1/PROFILE.db` folder where you stored your PROFILE.db file. You can list multiple profiles with this command\
> `-o` output folder\
> `-c` contig.db file\
> `--enforce-hierarchical-clustering` needed for later on

When you run``anvi-merge``:
* It will merge everything and create a merged profile
* It will attempt to create multiple clusterings of your splits using the default clustering configurations. 

Now we can perform genome **binning**. 
*"Basically, in metagenomic binning, you’re trying to group together a bunch of contigs that all belong to the same genome using various metrics like tetranucleotide frequency, differential coverage, completion, etc “ (https://anvio.org/help/7/artifacts/bin/).*

Here you are going to use two binners [Metabat2](https://bitbucket.org/berkeleylab/metabat/src/master/ ) and [MaxBin2](https://sourceforge.net/projects/maxbin2/).



#### Binning with Metabat2

```ssh
anvi-cluster-contigs -p ? -c ? -C METABAT --driver metabat2 --just-do-it --log-file log-metabat2

anvi-summarize -p /PATH/TO/merged_profiles/? -c ? -o SUMMARY_METABAT -C ?
```

<details><summary><b>Finished commands</b></summary>


```ssh
anvi-cluster-contigs -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db -C METABAT --driver metabat2 --just-do-it --log-file log-metabat2

anvi-summarize -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db -o SUMMARY_METABAT -C METABAT
```
</details>


> `-p` output folder from the last step/PROFILE.db file with the merged profiles\
> `-c` contig.db folder\
> `-C` name of the output collection that this step will create Note that this will be stored within your merged profiles and not as an extra file\
> `--driver` here you can name the driver you want to use\
> `--just-do-it` you need to specify this flag as `anvi-cluster-contigs` is an experimental workflow of the anvi´o program and therefore still under development. This way the developers want to make sure you are aware of it. 

#### Binning with MaxBin2
```ssh
anvi-cluster-contigs -p ? -c ? -C MAXBIN2 --driver maxbin2 --just-do-it --log-file log-maxbin2

anvi-summarize -p /PATH/TO/merged_profiles/? -c ? -o SUMMARY_MAXBIN2 -C ?
```

<details><summary><b>Finished commands</b></summary>
  
```ssh
anvi-cluster-contigs -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db -C MAXBIN2 --driver maxbin2 --just-do-it --log-file log-maxbin2

anvi-summarize -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db -o SUMMARY_MAXBIN2 -C MAXBIN2
```

</details>

### Questions
* Number of ${\color{red}ARCHAEA}$ bins you got from MetaBAT2? 
* Number of ${\color{red}ARCHAEA}$ bins you got from Maxbin2? 

> INSERT\
> YOUR\
> ANSWER\
> HERE

- we will use the BINs generated from METABAT2

## MAGs Quality Estimation

Estimate your genomes completeness and contamination levels.\
You can assess the quality of your bins by using 
- otherwise data is already available in the html file generated from binning
```
anvi-estimate-genome-completeness -c /PATH/TO/contigs.db -p /PATH/TO/merged_profiles/PROFILE.db -C METABAT
```
In the **next part** you will visualize and evaluate your results.\
If you want to check what collections you generated you can use:

```ssh
anvi-estimate-genome-completeness -p ? -c ? --list-collections
```

<details><summary><b>Finished commands</b></summary>

```ssh
anvi-estimate-genome-completeness -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db --list-collections
```

</details>

You can then use


```diff
-!!!!!!!!!!!!!!!!!!!!!AS MENTIONED BEFORE!!!!!!!!!!!!!!!!!!!!!
- Here you need to access anvi’o interactive -
- REPLACE the command line you want to run in interactive mode -
```

```
module load gcc12-env/12.1.0
module load miniconda3/4.12.0
conda activate anvio-8

anvi-interactive -p /PATH/TO/merged_profiles/PROFILE.db -c /PATH/TO/contigs.db -C YOUR_COLLECTION
```


``anvi-interactive`` gives you the possibility to manually inspect and work on bins.\
Once your browser window is open you can set all parameters that you want to set and ``click on draw``, in the bottom left corner. This will generate a graph that, depending on your settings, should look something like this:

![Pipeline]([anvi-interactive.jpg](https://github.com/AammarTufail/Biol217-2024/blob/main/anvi-interactive.jpg))

### Questions
* Which binning strategy gives you the best quality for the ${\color{red}ARCHAEA}$ bins?? 
* How many ${\color{red}ARCHAEA}$ bins do you get that are of High Quality? How many ${\color{red}BACTERIA}$ bins do you get that are of High Quality? 

> INSERT\
> YOUR\
> ANSWER\
> HERE

